Add a porcelain api for Github data

decentralion opened the pull request:
Interacting with raw contribution graphs is cumbersome. We'll need
more fluent and convenient ways to retrieve data from them; we can do
this by creating porcelain APIs that wrap the underlying graph.

This commit adds a simple porcelain API for the GitHub data. It creates
the following classes:

* `api.Repository`
* `api.Issue`
* `api.PullRequest`
* `api.Comment`
* `api.Author`

The classes all wrap a graph and a nodeAddress. They provide read-only
functions for retreiving data from the graph; that data might be a part
of the node payload, or it might do some graph traversal under the hood.

The choice to have the wrapper hold onto the Address rather than the
node itself was deliberate; in the future, the graph may contain nodes
that are not synchronously reachable, so this approach allows us to
create wrappers for nodes we can't synchronously reach. When this comes
up in practice, we can then add async methods to the wrapper.

Note that some data already included in our graph, such as
PullRequestReviews and PullRequestReviewComments, were deliberately
excluded, so as to allow the core ideas to be reviewed without
unnecessary clutter.

Test plan:
Check that the unit tests appropriately test the behavior, and that the
API seems pleasant to use.

wchargin commented on src/plugins/github/api.test.js:
```diff
@@ -0,0 +1,76 @@
+    expect(issue.url()).toBe(
+      "https://github.com/sourcecred/example-repo/issues/1"
+    );
+    expect(issue.type()).toMatchSnapshot();
```
Why is this not necessarily `"ISSUE"`?

decentralion replied:
Meant to test address - fixed.

wchargin commented on src/plugins/github/api.test.js:
```diff
@@ -0,0 +1,76 @@
+      "https://github.com/sourcecred/example-repo/issues/1"
+    );
+    expect(issue.type()).toMatchSnapshot();
+    expect(issue.authors()[0].login()).toBe("decentralion");
```
Stronger with `expect(issue.authors().map((x) => x.login())).toEqual(["decentralion"])`; up to you whether to change.

decentralion replied:
Done

wchargin commented on src/plugins/github/api.test.js:
```diff
@@ -0,0 +1,76 @@
+    );
+    expect(issue.type()).toMatchSnapshot();
+    expect(issue.authors()[0].login()).toBe("decentralion");
+  });
```
Did you want to test the `issue.address()`, as with the other object types?

decentralion replied:
Fixed.

wchargin commented on src/plugins/github/api.js:
```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+class GithubEntity<T: NodePayload> {
```
It is somewhat strange that a repository is not a `GithubEntity`. I suppose that this is because it is not actually a node in the graph… should it be? (No change requested.)

decentralion replied:
Having a repository node makes a lot of sense; we're almost sure to need them. Made an issue to track this suggestion. https://github.com/sourcecred/sourcecred/issues/171

decentralion commented:
Addressed your comments; also, per offline discussion, we now snapshot test the node and programatically verify that `Entity.address()` returns the node address.

Re your suggestion for composing the classes via higher-order functions: It's interesting, I think you'll agree we don't need the complexity at this time, but I'll keep it in mind for later.
